server/src/main/java/org/opensearch/search/aggregations/ConcurrentAggregationProcessor.java

```diff
@@ -32,53 +32,61 @@ public class ConcurrentAggregationProcessor extends DefaultAggregationProcessor
 
     @Override
     public void preProcess(SearchContext context) {
-        try {
-            if (context.aggregations() != null) {
-                if (context.aggregations().factories().hasNonGlobalAggregator()) {
-                    context.queryCollectorManagers().put(NonGlobalAggCollectorManager.class, new NonGlobalAggCollectorManager(context));
-                }
-                // initialize global aggregators as well, such that any failure to initialize can be caught before executing the request
-                if (context.aggregations().factories().hasGlobalAggregator()) {
-                    context.queryCollectorManagers().put(GlobalAggCollectorManager.class, new GlobalAggCollectorManager(context));
+        if (context.concurrentSegmentSearchEnabled()) {
```

Comment:
I think this no the way we should approach that:
 - the entry point for concurrent / non-concurrent / etc search paths is `QueryPhaseSearcher` 
 - right now `QueryPhaseSearcher` could 
     a) come from the plugin - we should respect that
     b) come from the concurrent search feature flag 
     c) come from default

To satisfy the (now) dynamic nature of the query phase, it would make sense to introduce the `QueryPhaseSearcher` wrapper which would make the decision which path to take, it could be initialized in the `SearchModule` with respect to the case(s) from above.

Reply:
@reta The idea here is when concurrent search will be GA'd then we can use the `ConcurrentQueryPhaseSearcher` as the default `QueryPhaseSearcher` unless one is provided by the plugin. However, we want to enable the concurrent path selectively based on cluster (global) and index (selective indices like remote vs doc rep) settings. This is to help users:
1. To try out their workload on concurrent path and see if that is best suited vs non concurrent path
2. If there are any regressions in concurrent path then fallback to non concurrent path dynamically

With the wrapper approach basically a wrapper will initialize both `Default` and `ConcurrentQueryPhaseSearcher`. Also return a wrapper for `AggregationProcessor` which can again have both `Default` and `ConcurrentAggregationProcessor`. Then based on the flag in the context, it can choose to invoke the methods of either default or concurrent implementation. With current mechanism, `ConcurrentQueryPhaseSearcher` and `ConcurrentAggregationProcessor` is acting as the wrapper instead. 

If a plugin is providing the `QueryPhaseSearcher` then that will take precedence over the wrapper and it will be upto the plugin implementation to decide on how to handle these flags ? Was thinking that usually a plugin will either derive from `Default` or `Concurrent` QueryPhaseSearcher and if it wants to control the behavior of the request to again either take concurrent vs non concurrent path, it will work as expected with current mechanism. With wrapper introduced, plugin implementation will need to be aware of the wrapper and derive from that ?

Reply:
> With current mechanism, ConcurrentQueryPhaseSearcher and ConcurrentAggregationProcessor is acting as the wrapper instead.

The `ConcurrentQueryPhaseSearcher` should not care about any index / clusters  settings or feature flags. It is supposed to implement the search + aggregation logic related to concurrent flow. If that is not enabled - it should never be invoked. 

> With the wrapper approach basically a wrapper will initialize both Default and ConcurrentQueryPhaseSearcher. Also return a wrapper for AggregationProcessor which can again have both Default and ConcurrentAggregationProcessor. 

Yes, in worst case wrapper may need to initialize 2 instances of the `QueryPhaseSearcher`, but this would be the only place where the decision which path to take is made. The `AggregationProcessor` is not a concern - it is coming from the `QueryPhaseSearcher`.

> If a plugin is providing the QueryPhaseSearcher then that will take precedence over the wrapper and it will be upto the plugin implementation to decide on how to handle these flags ?

Yes, if the plugin basically provide own implementation, we shouldn't intervene with it. 

> With wrapper introduced, plugin implementation will need to be aware of the wrapper and derive from that ?

The wrapper should be internal to OpenSearch core. The plugin authors have complete freedom on `QueryPhaseSearcher` implementation.

Reply:
Correct, the wrapper would help to get rid of obscure conditional logic so individual implementation would proceed cleanly.